src/main/java/com/google/devtools/build/lib/skyframe/FilesystemValueChecker.java

```diff
@@ -465,6 +467,19 @@ private boolean artifactIsDirtyWithDirectSystemCalls(
     try {
       FileArtifactValue fileMetadata =
           ActionMetadataHandler.fileArtifactValueFromArtifact(file, null, tsgm);
+
+      if (lastKnownData.isRemote() && lastKnownData.getType() == FileStateType.REGULAR_FILE
+          && !fileMetadata.isRemote() && fileMetadata.getType() == FileStateType.REGULAR_FILE) {
+        // If the file was remote but is now staged, we compare the digests.
+        byte[] digest = DigestUtils.manuallyComputeDigest(file.getPath(), fileMetadata.getSize());
+        if (Arrays.equals(digest, lastKnownData.getDigest())
+            && fileMetadata.getSize() == lastKnownData.getSize()) {
+          // TODO(chiwang): Find a way to update lastKnownData to fileMetadata so we don't need to
```

Comment:
One way to do this could be to attach a FileContentsProxy to a RemoteFileArtifactValue when we download the input. I think whether that idea makes sense would be informed by how we download the inputs -- how do we avoid re-downloading inputs today? By re-downloading, I mean a case of Action B depends on Action A and say we change something with Action B, but not A -- we would want not to materialize remote outputs of Action A again.

If we did that, we should see if the extra pointer takes more memory (for non-RBE case, we already have the actionId we don't use). Then, this code may not need any changes -- all you would need is to allow that content proxy comparison to work between RegularFileArtifactValue and RemoteFileArtifactValue.

Reply:
> how do we avoid re-downloading inputs today?

We guard the download with `FileArtifactValue.isRemote`.

After changing Action B, the following build will check Action A via action cache during which the RegularFileArtifactValues of staged outputs are generated. If Action A hit the cache, we will use these RegularFileArtifactValues for Action B. So there is no extra downloads for Action B in this run.

Reply:
> how do we avoid re-downloading inputs today?

Imagine we have action A executed remotely and action B locally.
 
1. Build 1 -- we run A remotely, download outputs of A to run action B.
2. We change a source of Action B so that it needs to be rerun.
2. Build 2 -- we run `bazel build :b` -- for A, we have a Skyframe cache hit, B we execute.

For A, we get an `ActionExecutionValue` from Skyframe, no action cache checks needed, hence the metadata for all its outputs should stay as remote. Will we download the inputs again in this case even though they are already available locally from prior build?

What happens if I change the downloaded file on disk between bazel builds?

Reply:
> Will we download the inputs again in this case even though they are already available locally from prior build?


In theory, we would download the inputs again because the ActionExecutionValue for Action A contains RemoteFileArtifactValue. However, in practice, and I have done some local tests, for some reason, Action A is invalidated and we check the action cache against it so the metadata becomes RegularFileArtifactValue when we run Action B. I don't totally understand the reason why Action A is invalidated but maybe related to the original issue.


> What happens if I change the downloaded file on disk between bazel builds?


The generating action (in this case, Action A) is invalidated and re-executed. Before execution, the downloaded (and modified) file will be deleted. Action A is executed remotely and then we get the action result without downloading the outputs.
1. If the result of Action A is same as before, Action B won't be re-executed because of action cache.
2. Otherwise, execute Action B, download the updated outputs of Action A.

Reply:
That's surprising -- I would think one should be able to invalidate a top-action without needing to change the base one. I would be more than happy to try to help you offline with constructing an example doing that.